core/src/main/kotlin/com/amazon/opendistroforelasticsearch/alerting/core/model/HttpInput.kt

```diff
@@ -0,0 +1,160 @@
+/*
+ *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.alerting.core.model
+
+import org.apache.commons.validator.routines.UrlValidator
+import org.apache.http.client.utils.URIBuilder
+import org.elasticsearch.common.CheckedFunction
+import org.elasticsearch.common.ParseField
+import org.elasticsearch.common.Strings
+import org.elasticsearch.common.xcontent.NamedXContentRegistry
+import org.elasticsearch.common.xcontent.ToXContent
+import org.elasticsearch.common.xcontent.XContentBuilder
+import org.elasticsearch.common.xcontent.XContentParser
+import org.elasticsearch.common.xcontent.XContentParserUtils
+import java.io.IOException
+
+/**
+ * This is a data class of HTTP type of input for Monitors.
+ */
+data class HttpInput(
+    val scheme: String,
+    val host: String?,
+    val port: Int,
+    val path: String?,
+    val params: Map<String, String>,
+    val url: String,
+    val connection_timeout: Int,
+    val socket_timeout: Int
+) : Input {
+
+    // Verify parameters are valid during creation
+    init {
+        require(validateFields()) {
+            "Either one of url or scheme + host + port+ + path + params can be set."
+        }
+        require(connection_timeout in 1..5) {
+            "Connection timeout: $connection_timeout is not in the range of 1 - 5"
+        }
+        require(socket_timeout in 1..60) {
+            "Socket timeout: $socket_timeout is not in the range of 1 - 60"
+        }
+
+        // Create an UrlValidator that only accepts "http" and "https" as valid scheme and allows local URLs.
+        val urlValidator = UrlValidator(arrayOf("http", "https"), UrlValidator.ALLOW_LOCAL_URLS)
+
+        // Build url field by field if not provided as whole.
+        val constructedUrl = if (Strings.isEmpty(url)) {
+            val uriBuilder = URIBuilder()
+            uriBuilder.scheme = scheme
+            uriBuilder.host = host
+            uriBuilder.port = port
+            uriBuilder.path = path
+            for (e in params.entries)
+                uriBuilder.addParameter(e.key, e.value)
+            uriBuilder.build()
+        } else {
+            URIBuilder(url).build()
+        }
+
+        // Make sure that when host is "localhost", only port 9200 is allowed.
+        if (constructedUrl.host == "localhost") {
+            require(constructedUrl.port == 9200) {
```

Comment:
We should check the current ES `http.port` setting to validate the port. In order to do this we need the `Settings` object which might be tricky to get in here.

If we cannot get `Settings` object here , one possible alternative I can think is to allow any port number in the Model but we can check during the runtime in `MonitorRunner`.

Reply:
Moved port validation from `HttpInput` to `RestIndexMonitorAction`.  Using the `http.port` from `Settings` to validate.